File: ui/conditions.js
'use strict';
// Reusable conditions are copied into a tournament at creation. Editing the
// library never changes an existing tournament or any recorded attempt.
loadPresets=async function(){
  state.presets=await api('presets');
  $('#presetList').innerHTML=state.presets.map((p,i)=>`<article class="preset"><h3>${esc(p.name)}</h3><p>${esc(tcText(p.settings.time_control))}<br>${p.settings.threads} thread(s) · ${p.settings.hash} MB hash<br>${p.settings.paired?'Paired openings':'Single games'} · Ponder ${p.settings.ponder?'ON':'OFF'}</p><button data-edit-preset="${i}">Edit</button> <button data-copy-preset="${i}">Save a copy</button></article>`).join('');
  $$('[data-edit-preset]').forEach(b=>b.onclick=()=>editPreset(state.presets[Number(b.dataset.editPreset)]));
  $$('[data-copy-preset]').forEach(b=>b.onclick=()=>editPreset(state.presets[Number(b.dataset.copyPreset)],true));
};
function editPreset(p=null,copy=false){
  const s=p?.settings||{threads:1,hash:1024,paired:true,ponder:false};
  modal(copy?'Copy test conditions':p?'Edit test conditions':'New test conditions',`${field('presetName','Preset name',p?p.name+(copy?' (copy)':''):'My control')}${tcEditor('ptc',s.time_control)}<div class="fields">${field('pThreads','Threads per engine',s.threads,'number','min="1" step="1"')}${field('pHash','Hash per engine (MB)',s.hash,'number','min="1" step="1"')}<label><input id="pPaired" type="checkbox" ${s.paired?'checked':''}> Paired openings</label><label><input id="pPonder" type="checkbox" ${s.ponder?'checked':''}> Enable pondering</label></div><p class="caption">Existing tournaments retain their recorded conditions. A different name saves another preset and keeps the original.</p>`,`<button class="primary" id="saveP">Save preset</button>`);
  wireTc('ptc');$('#saveP').onclick=safely(async()=>{
    await api('presets',{name:$('#presetName').value,original_name:p&&!copy?p.name:null,settings:{...s,time_control:readTc('ptc'),threads:Number($('#pThreads').value),hash:Number($('#pHash').value),paired:$('#pPaired').checked,ponder:$('#pPonder').checked}});
    $('#modal').close();await loadPresets();notify('Test conditions saved.');
  });
}
$('#savePreset').textContent='＋ New preset';$('#savePreset').onclick=()=>editPreset();
$('#presets').insertAdjacentHTML('beforeend',`<div class="panel position-library"><div class="panel-title"><h2>Starting positions</h2><button id="newPosition">＋ Set up a position</button></div><div id="positionList"></div><p class="caption">Place pieces on the board or paste a FEN. Saved positions can be selected when creating a tournament.</p></div>`);
const loadConditionsPresets=loadPresets;
loadPresets=async function(){await loadConditionsPresets();await loadPositions();};
async function loadPositions(){
  const items=await api('positions');
  $('#positionList').innerHTML=items.length?items.map((p,i)=>`<article class="preset"><h3>${esc(p.name)}</h3><p>${p.chess960?'Chess960':'Standard chess'}</p><code>${esc(p.fen)}</code><div class="position-actions"><button data-edit-position="${i}">Edit position</button><button data-export-position="${i}">Export FEN</button></div></article>`).join(''):'<p class="caption">No saved positions yet.</p>';
  $$('[data-edit-position]').forEach(b=>b.onclick=()=>editPosition(items[Number(b.dataset.editPosition)]));
  $$('[data-export-position]').forEach(b=>b.onclick=()=>download('starting-position.fen',items[Number(b.dataset.exportPosition)].fen+'\n','text/plain'));
}
$('#newPosition').onclick=()=>editPosition();
function editPosition(saved=null){
  const initial='rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1';
  let cells=[],piece='P',flip=false,revision=0,timer;
  modal('Set up a starting position',`<div class="position-editor"><div><div id="setupBoard"></div><div class="toolbar position-actions"><button id="setupFlip">Flip board</button><button id="setupInitial">Initial position</button><button id="setupClear">Clear board</button></div><div id="piecePalette" class="piece-palette">${'KQRBNPkqrbnp'.split('').map(p=>`<button type="button" data-place-piece="${p}" aria-label="Place ${p===p.toUpperCase()?'White':'Black'} ${{K:'king',Q:'queen',R:'rook',B:'bishop',N:'knight',P:'pawn'}[p.toUpperCase()]}"><img alt="" src="/pieces/${state.pieceSet||'classic'}/${p===p.toUpperCase()?'w':'b'}${p.toUpperCase()}.svg"></button>`).join('')}<button data-place-piece="" class="erase-piece">Erase</button></div><p class="caption">Choose a piece, then click a square. Right-click to erase. Squares also support Tab and Enter.</p></div><div>${field('positionName','Position name',saved?.name||'My starting position')}<label class="setup-fen">FEN<textarea id="positionFen" rows="3" spellcheck="false"></textarea></label><div class="fields"><label>Side to move<select id="setupTurn"><option value="w">White</option><option value="b">Black</option></select></label>${field('setupCastle','Castling rights (KQkq or rook files for Chess960)','KQkq')}${field('setupEp','En-passant square (or -)','-')}${field('setupHalf','Halfmove clock',0,'number','min="0" step="1"')}${field('setupFull','Fullmove number',1,'number','min="1" step="1"')}<label><input type="checkbox" id="setup960" ${saved?.chess960?'checked':''}> Chess960 rules</label></div><div class="position-actions">${field('setup960Index','Chess960 start (0–959)',518,'number','min="0" max="959" step="1"')}<button id="setupGenerate">Generate start</button></div><p id="positionValidation" role="status"></p></div></div>`,`<button id="savePosition" class="primary" disabled>Save position</button>`);
  const fenNode=$('#positionFen'),statusNode=$('#positionValidation'),saveNode=$('#savePosition');
  const alive=()=>$('#positionFen')===fenNode&&$('#modal').open;
  const validate=async()=>{
    const current=++revision;saveNode.disabled=true;statusNode.textContent='Checking position…';
    try{const p=await api('position-preview',{fen:fenNode.value,chess960:$('#setup960').checked});
      if(current!==revision||!alive())return;
      statusNode.textContent=p.valid?`Valid position · ${p.legal_moves.length} legal moves${p.check?' · in check':''}`:p.errors.join(' · ');
      statusNode.className=p.valid?'position-valid':'pairing-error';saveNode.disabled=!p.valid;
    }catch(e){if(current===revision&&alive()){statusNode.textContent=e.message;statusNode.className='pairing-error';}}
  };
  const schedule=()=>{++revision;clearTimeout(timer);saveNode.disabled=true;timer=setTimeout(()=>{if(alive())validate();},160);};
  const placement=()=>Array.from({length:8},(_,r)=>{let row='',empty=0;for(const p of cells.slice(r*8,r*8+8)){if(!p)empty++;else{if(empty)row+=empty;empty=0;row+=p;}}return row+(empty||'');}).join('/');
  const write=()=>{fenNode.value=`${placement()} ${$('#setupTurn').value} ${$('#setupCastle').value.trim()||'-'} ${$('#setupEp').value.trim()||'-'} ${$('#setupHalf').value} ${$('#setupFull').value}`;draw();schedule();};
  const draw=()=>{
    $('#setupBoard').innerHTML=boardHtml(placement(),'',flip);
    $$('#setupBoard [data-square]').forEach(el=>{
      const index=(8-Number(el.dataset.square[1]))*8+'abcdefgh'.indexOf(el.dataset.square[0]);
      el.setAttribute('role','button');el.tabIndex=0;el.setAttribute('aria-label',`${el.dataset.square}, ${el.querySelector('img')?.alt||'empty'}`);
      const put=p=>{cells[index]=p;write();$('#setupBoard [data-square="'+el.dataset.square+'"]').focus({preventScroll:true});};
      el.onclick=()=>put(piece);el.oncontextmenu=e=>{e.preventDefault();put('');};el.onkeydown=e=>{if(['Enter',' '].includes(e.key)){e.preventDefault();put(piece);}};
    });
  };
  const read=()=>{
    try{
      const parts=fenNode.value.trim().split(/\s+/),rows=parts[0].split('/'),next=[];
      if(rows.length!==8)throw Error('FEN needs eight ranks.');
      for(const row of rows){const start=next.length;for(const c of row){if(/[1-8]/.test(c))next.push(...Array(Number(c)).fill(''));else if(/[KQRBNPkqrbnp]/.test(c))next.push(c);else throw Error('Invalid FEN piece.');}if(next.length-start!==8)throw Error('Each FEN rank needs eight squares.');}
      cells=next;$('#setupTurn').value=parts[1]||'w';$('#setupCastle').value=parts[2]||'-';$('#setupEp').value=parts[3]||'-';$('#setupHalf').value=parts[4]??0;$('#setupFull').value=parts[5]??1;draw();schedule();
    }catch(e){++revision;clearTimeout(timer);saveNode.disabled=true;statusNode.textContent=e.message;statusNode.className='pairing-error';}
  };
  fenNode.value=saved?.fen||initial;fenNode.oninput=read;
  $$('#piecePalette button').forEach(b=>b.onclick=()=>{piece=b.dataset.placePiece;$$('#piecePalette button').forEach(x=>x.classList.toggle('primary',x===b));});
  $('#piecePalette [data-place-piece="P"]').classList.add('primary');
  ['setupTurn','setupCastle','setupEp','setupHalf','setupFull','setup960'].forEach(id=>$('#'+id).onchange=write);
  $('#setupFlip').onclick=()=>{flip=!flip;draw();};
  $('#setupInitial').onclick=()=>{$('#setup960').checked=false;fenNode.value=initial;read();};
  $('#setupClear').onclick=()=>{fenNode.value='8/8/8/8/8/8/8/8 w - - 0 1';read();};
  $('#setupGenerate').onclick=safely(async()=>{
    const current=++revision;clearTimeout(timer);saveNode.disabled=true;
    const p=await api('position-preview',{chess960_index:Number($('#setup960Index').value)});
    if(current!==revision||!alive())return;$('#setup960').checked=true;fenNode.value=p.fen;read();
  });
  saveNode.onclick=safely(async()=>{
    await api('positions',{name:$('#positionName').value,fen:fenNode.value,chess960:$('#setup960').checked,original_name:saved?.name||null});
    ++revision;clearTimeout(timer);$('#modal').close();await loadPositions();notify('Starting position saved. Existing tournaments keep their recorded opening.');
  });
  read();
}
function startingPositionPayload(){
  const source=$('#topeningSource')?.value||'initial';
  return {saved_position:source==='saved'?$('#tsavedPosition').value:null,opening_file:source==='file'?$('#topening').value:'',all960:source==='all960'};
}
const tournamentBeforeConditions=newTournament;
newTournament=async function(){
  await tournamentBeforeConditions();const node=$('#tname');const positions=await api('positions');if($('#tname')!==node||!$('#modal').open)return;
  const openings=$('#topening').closest('.fields');
  openings.insertAdjacentHTML('beforebegin',`<div class="fields opening-selection"><label>Starting position source<select id="topeningSource"><option value="initial">Standard initial position</option><option value="file">Opening file or Polyglot book</option><option value="saved" ${positions.length?'':'disabled'}>Saved starting position</option><option value="all960">All 960 Chess960 starts</option></select></label><label id="savedPositionField" hidden>Saved position<select id="tsavedPosition">${positions.map(p=>`<option value="${esc(p.name)}">${esc(p.name)} · ${p.chess960?'Chess960':'Standard'}</option>`).join('')}</select></label><label id="file960Field" hidden><input id="tfile960" type="checkbox"> Opening file uses Chess960 rules</label></div>`);
  $('#t960').closest('label').hidden=true;
  const sourceChanged=()=>{const s=$('#topeningSource').value;$('#savedPositionField').hidden=s!=='saved';$('#file960Field').hidden=s!=='file';$('#topening').closest('.field').hidden=s!=='file';$('#tdepth').closest('.field').hidden=s!=='file';};
  $('#topeningSource').onchange=sourceChanged;sourceChanged();
  const presetChanged=$('#tpreset').onchange;$('#tpreset').onchange=()=>{const key=$('#tpreset').value;if(key!=='')$('#tponder').checked=state.presets[Number(key)].settings.ponder;return presetChanged?.();};
};
$('#newTournament').onclick=safely(newTournament);

File: arena/conditions.py
"""Validated reusable conditions and editable starting positions."""
from dataclasses import asdict
import json
import chess
from .models import TimeControl

def preset_settings(value):
    if not isinstance(value,dict):raise ValueError('Preset settings must be an object')
    result=dict(value);result['time_control']=asdict(TimeControl.parse(value.get('time_control')))
    for key,default in (('threads',1),('hash',1024)):
        n=value.get(key,default)
        if not isinstance(n,int) or isinstance(n,bool) or n<1:raise ValueError(key+' must be a positive integer')
        result[key]=n
    for key,default in (('paired',True),('ponder',False)):
        if not isinstance(value.get(key,default),bool):raise ValueError(key+' must be true or false')
        result[key]=value.get(key,default)
    return result

def save_preset(store,name,value,original=None):
    name=str(name).strip()
    if not name:raise ValueError('Enter a preset name')
    result=preset_settings(value)
    with store.tx():
        if original is not None and not store.one('SELECT name FROM presets WHERE name=?',(original,)):raise ValueError('Original preset no longer exists')
        if name!=original and store.one('SELECT name FROM presets WHERE name=?',(name,)):raise ValueError('That name already exists. Edit that preset or choose another name.')
        store.db.execute('INSERT OR REPLACE INTO presets VALUES(?,?)',(name,json.dumps(result)))
        # Renaming creates a new name while retaining the original as a reusable
        # copy; existing tournament snapshots never change.
    return {'name':name,'settings':result}

def position(value):
    if not isinstance(value,dict):raise ValueError('Position must be an object')
    variant=value.get('chess960',False)
    if not isinstance(variant,bool):raise ValueError('Chess960 must be true or false')
    if 'chess960_index' in value:
        n=value['chess960_index']
        if not isinstance(n,int) or isinstance(n,bool) or not 0<=n<960:raise ValueError('Chess960 start number must be 0–959')
        board=chess.Board.from_chess960_pos(n);variant=True
    else:
        fen=value.get('fen',chess.STARTING_FEN)
        if not isinstance(fen,str):raise ValueError('FEN must be text')
        fields=fen.split()
        if len(fields)!=6:raise ValueError('FEN must contain all six fields')
        if not fields[4].isdigit() or not fields[5].isdigit() or int(fields[5])<1:raise ValueError('Halfmove clock must be non-negative; fullmove number must be positive')
        board=chess.Board(fen,chess960=variant)
    status=board.status()
    errors=[label for flag,label in (
        (chess.STATUS_NO_WHITE_KING,'White king is missing'),(chess.STATUS_NO_BLACK_KING,'Black king is missing'),
        (chess.STATUS_TOO_MANY_KINGS,'Too many kings'),(chess.STATUS_TOO_MANY_WHITE_PAWNS,'Too many White pawns'),
        (chess.STATUS_TOO_MANY_BLACK_PAWNS,'Too many Black pawns'),(chess.STATUS_PAWNS_ON_BACKRANK,'Pawns on a back rank'),
        (chess.STATUS_TOO_MANY_WHITE_PIECES,'Too many White pieces'),(chess.STATUS_TOO_MANY_BLACK_PIECES,'Too many Black pieces'),
        (chess.STATUS_BAD_CASTLING_RIGHTS,'Castling rights do not match the king and rook squares'),
        (chess.STATUS_INVALID_EP_SQUARE,'Invalid en-passant square'),(chess.STATUS_OPPOSITE_CHECK,'The side that just moved is in check'),
        (chess.STATUS_TOO_MANY_CHECKERS,'Too many checking pieces'),(chess.STATUS_IMPOSSIBLE_CHECK,'Impossible check arrangement'),
        (chess.STATUS_EMPTY,'The board is empty')) if status&flag]
    return {'fen':board.fen(shredder=variant,en_passant='fen'),'chess960':variant,'valid':board.is_valid(),'errors':errors,
      'legal_moves':[{'uci':board.uci(m),'san':board.san(m)} for m in board.legal_moves] if board.is_valid() else [],'check':board.is_check()}

def save_position(store,name,value):
    name=str(name).strip()
    if not name:raise ValueError('Enter a position name')
    checked=position(value)
    if not checked['valid']:raise ValueError('; '.join(checked['errors']) or 'Invalid position')
    item={'name':name,'fen':checked['fen'],'chess960':checked['chess960']}
    original=value.get('original_name')
    with store.tx():
        if original is not None and not store.one('SELECT name FROM positions WHERE name=?',(original,)):raise ValueError('Original position no longer exists')
        if name!=original and store.one('SELECT name FROM positions WHERE name=?',(name,)):raise ValueError('That position name already exists. Edit it or choose another name.')
        store.db.execute('INSERT OR REPLACE INTO positions VALUES(?,?)',(name,json.dumps(item)))
    return item
